src/main/java/com/mcmoddev/communitymod/SubModLoader.java

```diff
@@ -0,0 +1,217 @@
+                if (!data.getAnnotationInfo().containsKey("clientSideOnly") || (boolean) data.getAnnotationInfo().get("clientSideOnly") && event.getSide().isClient())
+                {
+                    ISubMod subMod = (ISubMod) Class.forName(data.getClassName()).newInstance();
+                    String description = (String) data.getAnnotationInfo().getOrDefault("description", "No description is provided for this submod.");
+                    String attribution = (String) data.getAnnotationInfo().getOrDefault("attribution", "No attribution is provided for this submod.");
+                    Boolean requiresMcRestart = (Boolean) data.getAnnotationInfo().getOrDefault("requiresMcRestart", true);
+                    SubModContainer container = new SubModContainer(modName, description, attribution, subMod, requiresMcRestart);
+
+                    if (subMod.getClass().isAnnotationPresent(Mod.EventBusSubscriber.class))
+                    {
+                        eventSubscriberWarns.add(container);
+                        MinecraftForge.EVENT_BUS.unregister(subMod.getClass());
+                    }
+
+                    if (CommunityMod.getConfig().isSubModEnabled(container))
+                    {
+                        loadedSubMods.add(container);
+                        MinecraftForge.EVENT_BUS.register(subMod.getClass());
+
+                        LOGGER.info("Loaded submod {} from class {}.", modName, data.getClassName());
+                    }
+                    else if (!requiresMcRestart)
+                    {
+                        MinecraftForge.EVENT_BUS.register(subMod.getClass());
+                    }
+
+                    subMods.put(subMod, container);
+                }
+            }
+            catch (Exception e)
+            {
+                LOGGER.error("Failed to load submod {}.", modName);
+                LOGGER.catching(e);
+            }
+        }
+
+        for (SubModContainer container : eventSubscriberWarns)
+        {
+            LOGGER.warn("{} from {} uses an @EventBusSubscriber annotation, which is normally done automagically. Proceeding anyway.", container.getSubMod().getClass(), container.getName());
+        }
+
+        LOGGER.info("Loaded {} submods in {}ms.", loadedSubMods.size(), System.currentTimeMillis() - startTime);
+
+        for (SubModContainer container : loadedSubMods)
+        {
+            container.getSubMod().onConstruction(event);
+        }
+    }
+
+    public static List<SubModContainer> getLoadedSubMods()
+    {
+        return Collections.unmodifiableList(INSTANCE.loadedSubMods);
+    }
+
+    public static Collection<SubModContainer> getSubMods()
+    {
+        return Collections.unmodifiableCollection(INSTANCE.subMods.values());
+    }
+
+    public static boolean isSubModLoaded(ISubMod subMod)
+    {
+        return INSTANCE.loadedSubMods.contains(INSTANCE.subMods.get(subMod));
+    }
+
+    public static boolean unload(ISubMod subMod)
```

Comment:
Event listeners of the sub mod (registered by default by this PR) are not unregistered when it is unloaded?

Reply:
Event handlers are dynamically generated using ASM by Forge, so unregistering one post-init wouldn't change anything.